src/compiler/types.ts

```diff
@@ -4968,7 +4968,16 @@ export interface TypeChecker {
     getWidenedLiteralType(type: Type): Type;
     /** @internal */
     getPromisedTypeOfPromise(promise: Type, errorNode?: Node): Type | undefined;
-    /** @internal */
+    /**
+     * Gets the "awaited type" of a type.
+     *
+     * The "awaited type" of an expression is its "promised type" if the expression is a
+     * Promise-like type; otherwise, it is the type of the expression. If the "promised
+     * type" is itself a Promise-like, the "promised type" is recursively unwrapped until a
+     * non-promise type is found.
+     *
+     * This is used to reflect the runtime behavior of the `await` keyword.
+     */

```

Comment:
The original comment contains a bit of "inside baseball" terminology that isn't explained elsewhere in the API documentation, such as what a "promised type" is. I would amend this comment to clarify:

```suggestion
    /**
     * Gets the "awaited type" of a type.
     *
     * If an expression has a Promise-like type, the "awaited type" of the expression is
     * derived from the type of the first argument of the fulfillment callback for that
     * Promise's `then` method. If the "awaited type" is itself a Promise-like, it is
     * recursively unwrapped until a non-promise type is found.
     *
     * If an expression does not have Promise-like type, its "awaited type" is the type
     * of the expression.
     *
     * In both cases, if the resulting  "awaited type" is a generic object type, then it
     * is wrapped in an `Awaited<T>`.
     *
     * This is used to reflect the runtime behavior of the `await` keyword.
     */
```

Reply:
When does `getAwaitedType` return `undefined`?

Reply:
(I think only in erroring cases but I think it'd would be worth documenting.)

Reply:
It only returns `undefined` in two cases: A promise that recursively unwraps to itself, or a non-Promise "thenable", which is an object with a callable `then()` method that doesn't accept a callback as its first argument.

Reply:
I've amended my suggestion above to explain the failure cases.